### ISSUE TITLE
reminder: index-wayback needs timestamps & tar-offsets for its operation

### ISSUE THREAD

**Issue by hvr** [well-typed/hackage-security#129]
In order to support the `--index-wayback` feature I'm working on, for all cache entries, I need timestamps & tar-offsets of **all** events in the 01-index.tar file. Specifically I modified the data type as follows in my patch:

```diff
-data IndexCacheEntry = CachePackageId PackageId BlockNo
+type Timestamp = Int64
+
+data IndexCacheEntry = CachePackageId PackageId BlockNo Timestamp
                      | CacheBuildTreeRef BuildTreeRefType BlockNo
-                     | CachePreference Dependency
+                     | CachePreference Dependency BlockNo Timestamp
   deriving (Eq)
```

See also https://github.com/hvr/cabal/tree/index-wayback

**Comment by dcoutts** [comment]
This is actually going to be tricky with the approach we've been following. I'm not inclined to switch right now since we have limited time to get this done. But it'd be fine to change later, especially if you can persuade someone to do it.

The problem is that we're using an index of the tar file to generate the cache, and that index already does the masking, later things masking earlier entries. To switch to having the cache contain everything (and doing the masking on cache read) we would need to not use the tar index but instead do a sequential scan.

**Comment by dcoutts** [comment]
Specifically we'd need to add an API to do a sequential scan, and then redo the cache in terms of that.

The other thing we'll have to deal with is changing the format of the cache file, thinking about sharing between multiple cabal-install versions.

**Comment by hvr** [comment]
> The other thing we'll have to deal with is changing the format of the cache file, thinking about sharing between multiple cabal-install versions.

Already thought about that long time ago: #117 :-)

**Comment by edsko** [comment]
So in discussion with @hvr I have now added an API for scanning through the index sequentially, and an example how to use it in the `example-client`. However, we have decided that

* We will keep the `00-index.cache` as it currently is. This means that `cabal-install` can use the index built by the security library to construct `00-index.cache`, as this index only contains the latest versions of all files.
* @hvr In his own branch will _additionally_ add a `01-index.cache` to `cabal`. This cache can be built lazily (i.e., when required, for instance by the way-back feature). This is outside the scope of the security project proper, but the new sequential access to the index can be used to build this cache.

So I'm closing both this issue and #117, as I think further support from the security library is not required.
